ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DataColumnSidecarRecoveringCustodyImpl.java

```diff
@@ -173,28 +174,34 @@ private boolean shouldSkipProcessing(final UInt64 slot) {
     return true;
   }
 
-  private synchronized void createOrUpdateRecoveryTaskForBlock(final BeaconBlock block) {
-    if (recoveryTasks.containsKey(block.getSlotAndBlockRoot())) {
-      final RecoveryTask existing = recoveryTasks.get(block.getSlotAndBlockRoot());
-      if (existing.block().get() == null) {
-        existing.block().set(block);
-        maybeStartRecovery(existing);
+  private void createOrUpdateRecoveryTaskForBlock(final BeaconBlock block) {
+    if (!recoveryTasks.containsKey(block.getSlotAndBlockRoot())) {
+      final RecoveryTask oldValue =
+          recoveryTasks.putIfAbsent(
+              block.getSlotAndBlockRoot(),
+              new RecoveryTask(
+                  new AtomicReference<>(block),
+                  new ConcurrentHashMap<>(),
+                  new AtomicBoolean(false),
+                  new AtomicBoolean(false)));
+      if (oldValue == null) {
+        return;
       }
-    } else {
-      recoveryTasks.put(
-          block.getSlotAndBlockRoot(),
-          new RecoveryTask(
-              new AtomicReference<>(block),
-              new ConcurrentHashSet<>(),
-              new AtomicBoolean(false),
-              new AtomicBoolean(false)));
+    }
+    final RecoveryTask existing = recoveryTasks.get(block.getSlotAndBlockRoot());
+    if (existing.block().get() == null) {
+      existing.block().set(block);
+      maybeStartRecovery(existing);
     }
   }
 
-  private synchronized void maybeStartRecovery(final RecoveryTask task) {
+  private void maybeStartRecovery(final RecoveryTask task) {
     if (readyToBeRecovered(task)) {
-      task.recoveryStarted().set(true);
-      if (task.existingColumnIds().size() != columnCount) {
+      if (task.recoveryStarted.compareAndSet(false, true)) {
+        if (task.existingSidecars.size() == columnCount) {
+          task.existingSidecars.clear();
+          return;
+        }
         asyncRunner.runAsync(() -> prepareAndInitiateRecovery(task)).finishError(LOG);
```

Comment:
so an `alwaysRun` at the end setting the recoveryStarted -> false for example (from the move comment about recoveryStarted) would ensure we can see the place we effectively unset the `recoveryStarted`